hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/utils/HoodieArrayWritableAvroUtils.java

```diff
@@ -19,20 +19,356 @@
+          return new DoubleWritable(((LongWritable) writable).get());
+        }
+        break;
+      case BYTES:
+        if (oldSchema.getType() == Schema.Type.STRING) {
+          return new BytesWritable(getUTF8Bytes(writable.toString()));
+        }
+        break;
+      case STRING:
+        if (oldSchema.getType() == Schema.Type.ENUM) {
+          return writable;
+        }
+        if (oldSchema.getType() == Schema.Type.BYTES) {
+          return new Text(StringUtils.fromUTF8Bytes(((BytesWritable) writable).getBytes()));
+        }
+        if (oldSchema.getLogicalType() == LogicalTypes.date()) {
+          return new Text(toJavaDate(((IntWritable) writable).get()).toString());
+        }
+        if (oldSchema.getType() == Schema.Type.INT
+            || oldSchema.getType() == Schema.Type.LONG
+            || oldSchema.getType() == Schema.Type.FLOAT
+            || oldSchema.getType() == Schema.Type.DOUBLE) {
+          return new Text(writable.toString());
+        }
+        if (oldSchema.getType() == Schema.Type.FIXED && oldSchema.getLogicalType() instanceof LogicalTypes.Decimal) {
+          HiveDecimalWritable hdw = (HiveDecimalWritable) writable;
+          return new Text(hdw.getHiveDecimal().bigDecimalValue().toPlainString());
+        }
+        break;
+      case FIXED:
+        if (newSchema.getLogicalType() instanceof LogicalTypes.Decimal) {
+          LogicalTypes.Decimal decimal = (LogicalTypes.Decimal) newSchema.getLogicalType();
+          DecimalTypeInfo decimalTypeInfo = new DecimalTypeInfo(decimal.getPrecision(), decimal.getScale());
+
+          if (oldSchema.getType() == Schema.Type.STRING
+              || oldSchema.getType() == Schema.Type.INT
+              || oldSchema.getType() == Schema.Type.LONG
+              || oldSchema.getType() == Schema.Type.FLOAT
+              || oldSchema.getType() == Schema.Type.DOUBLE) {
+            // loses trailing zeros due to hive issue. Since we only read with hive, I think this is fine
+            HiveDecimalWritable converted = new HiveDecimalWritable(HiveDecimal.create(new java.math.BigDecimal(writable.toString())));
+            return HiveDecimalUtils.enforcePrecisionScale(converted, decimalTypeInfo);
+          }
+
+          if (oldSchema.getType() == Schema.Type.BYTES) {
+            ByteBuffer buffer = ByteBuffer.wrap(((BytesWritable) writable).getBytes());
+            BigDecimal bd = new BigDecimal(new BigInteger(buffer.array()), decimal.getScale());
+            HiveDecimalWritable converted = new HiveDecimalWritable(HiveDecimal.create(bd));
+            return HiveDecimalUtils.enforcePrecisionScale(converted, decimalTypeInfo);
+          }
+        }
+        break;
+      default:
+    }
+    throw new HoodieAvroSchemaException(String.format("cannot support rewrite value for schema type: %s since the old schema type is: %s", newSchema, oldSchema));
+  }
+
+  private static final Cache<Schema, ArrayWritableObjectInspector> OBJECT_INSPECTOR_CACHE =
```

Comment:
Previously, the ObjectInspectorCache has this as an instance variable so the lifecycle of the cache is tied to that instance. Now this will live on for the life of the jvm it runs in. What is the reason for this change?